openml/runs/trace.py

```diff
@@ -26,56 +30,348 @@ def get_selected_iteration(self, fold, repeat):
+        for trace_idx in range(len(trace_arff['data'])):
+            # iterate over first three entrees of a trace row
+            # (fold, repeat, trace_iteration) these should be int
             for line_idx in range(3):
-                value = trace['data'][trace_idx][line_idx]
-                trace['data'][trace_idx][line_idx] = int(trace['data'][trace_idx][line_idx])
-        return trace
+                trace_arff['data'][trace_idx][line_idx] = int(
+                    trace_arff['data'][trace_idx][line_idx]
+                )
+
+        return cls.trace_from_arff(trace_arff)
+
+    def to_filesystem(self, file_path):
+        """Serialize the trace object to the filesystem.
+
+        Serialize the trace object as an arff.
+
+        Parameters
+        ----------
+        file_path: str
+            File path where the trace arff will be stored.
+        """
+
+        trace_arff = arff.dumps(self.trace_to_arff())
+        with open(os.path.join(file_path, 'trace.arff'), 'w') as f:
+            f.write(trace_arff)
+
+    def trace_to_arff(self):
+        """Generate the arff dictionary for uploading predictions to the server.
+
+        Uses the trace object to generate an arff dictionary representation.
+
+        Returns
+        -------
+        arff_dict : dict
+            Dictionary representation of the ARFF file that will be uploaded.
+            Contains information about the optimization trace.
+        """
+        if self.trace_iterations is None:
+            raise ValueError("trace_iterations missing from the trace object")
+
+        # attributes that will be in trace arff
+        trace_attributes = [
+            ('repeat', 'NUMERIC'),
+            ('fold', 'NUMERIC'),
+            ('iteration', 'NUMERIC'),
+            ('evaluation', 'NUMERIC'),
+            ('selected', ['true', 'false']),
+        ]
+        trace_attributes.extend([
+            ('parameter_' + parameter, 'STRING') for parameter in
+            next(iter(self.trace_iterations.values())).get_parameters()
+        ])
+
+        arff_dict = OrderedDict()
+        data = []
+        for trace_iteration in self.trace_iterations.values():
+            tit_list = []
+            for attr, _ in trace_attributes:
+                if attr.startswith('parameter_'):
+                    attr = attr[len('parameter_'):]
+                    value = trace_iteration.get_parameters()[attr]
+                else:
+                    value = getattr(trace_iteration, attr)
+                if attr == 'selected':
+                    if value:
+                        tit_list.append('true')
+                    else:
+                        tit_list.append('false')
+                else:
+                    tit_list.append(value)
+            data.append(tit_list)
+
+        arff_dict['attributes'] = trace_attributes
+        arff_dict['data'] = data
+        arff_dict['relation'] = "Trace"
```

Comment:
Potentially, we can also add the arff_dict description attribute to say something like "Auto-generated trace by search procedure XXX on model YYY", although I wouldn't know what the exact content should be. For now, leave a TODO?